File: src/leetcode/second/Code_148_SortList.java
package leetcode.second;

import leetcode.tool.LinkedListTool;
import leetcode.tool.ListNode;

/**
 * 在 O(n log n) 时间复杂度和常数级空间复杂度下，对链表进行排序。
 *
 * @since 2020-3-7
 */
public class Code_148_SortList {
    static ListNode sortList(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode fast = head, slow = head;
        while (fast.next != null && fast.next.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }
        ListNode mid = slow.next;
        slow.next = null;
        ListNode left = sortList(head);
        ListNode right = sortList(mid);
        return merge(left, right);
    }

    private static ListNode merge(ListNode left, ListNode right) {
        ListNode dummy = new ListNode(Integer.MIN_VALUE);
        ListNode cur = dummy;
        while (left != null && right != null) {
            if (left.val < right.val) {
                cur.next = left;
                left = left.next;
            } else {
                cur.next = right;
                right = right.next;
            }
            cur = cur.next;
        }
        cur.next = left == null ? right : left;
        return dummy.next;
    }

    public static void main(String[] args) {
        ListNode h1 = LinkedListTool.generateList(new int[]{4, 2, 1, 3});
        LinkedListTool.printList(sortList(h1));
        ListNode h2 = LinkedListTool.generateList(new int[]{6, 2, 2, 1, 3});
        LinkedListTool.printList(sortList(h2));
    }
}

File: src/cpp-leetcode/first/Code_148_SortList.cpp
#include <vector>

#include "../List.h"

/**
 * sort linked list in O(nlogn) time and O(1) space comple
 * merge sort
 *
 * @since 2020-9-27 Sunday 11:42 - 11:55
 */
class Solution {
   public:
    ListNode* sortList(ListNode* head) {
        if (head == nullptr || head->next == nullptr)
            return head;
        ListNode *f = head, *s = head;
        while (f->next != nullptr && f->next->next != nullptr) {
            f = f->next->next;
            s = s->next;
        }
        ListNode* mid = s->next;
        s->next = nullptr;
        ListNode* l = sortList(head);
        ListNode* r = sortList(mid);
        ListNode *dummy = new ListNode(-1), *cur = dummy;
        while (l != nullptr && r != nullptr) {
            if (l->val < r->val) {
                cur->next = l;
                l = l->next;
            } else {
                cur->next = r;
                r = r->next;
            }
            cur = cur->next;
        }
        if (l != nullptr) {
            cur->next = l;
        }
        if (r != nullptr) {
            cur->next = r;
        }
        return dummy->next;
    }
};

int main(int argc, char const* argv[]) {
    Solution s;
    std::vector<int> v = {3, 14, -1, 12, 0, 1};
    ListNode* h = List::linkedList(v);
    ListNode* res = s.sortList(h);
    List::printList(res);
}
